smart/model.py

```diff
@@ -982,32 +982,16 @@ def _init_4_7_set_initial_conditions(self):
                     )
                 if species.has_subdomain:
                     # restrict to specified subdomain
-                    uvec = self.cc[species.compartment_name].u["u"].vector()
-                    values = uvec.get_local()
-                    values_new = np.zeros_like(values)
-                    mesh_ref = self.parent_mesh.dolfin_mesh
-                    funcSpace = species.V
-                    bmesh = funcSpace.mesh()
-                    store_map = bmesh.topology().mapping()[mesh_ref.id()].vertex_map()
-                    idx_list = species.subdomain_data.where_equal(species.subdomain_val)
-                    for idx in idx_list:
-                        facet = d.Facet(mesh_ref, idx)
-                        for vertex in d.vertices(facet):
-                            global_idx = vertex.global_index()
-                            local_idx = np.nonzero(np.array(store_map) == global_idx)
-                            if len(local_idx[0]) == 0:
-                                continue
-                            else:
-                                local_idx = local_idx[0][0]
-                                cur_sub_idx = d.vertex_to_dof_map(funcSpace)[local_idx]
-                                values_new[species.dof_map[cur_sub_idx]] = values[
-                                    species.dof_map[cur_sub_idx]
-                                ]
-
+                    u_cur = self.cc[species.compartment_name].u[ukey]
+                    u_new = self.create_restriction(
+                        u_cur, species.subdomain_data, species.subdomain_val
+                    )
+                    u_vec = u_cur.vector()
+                    values = u_vec.get_local()
+                    values_new = u_new.vector().get_local()
                     values[species.dof_map] = values_new[species.dof_map]
-                    vec = self.cc[species.compartment_name].u[ukey].vector()
-                    vec.set_local(values)
-                    vec.apply("insert")
+                    u_vec.set_local(values)
+                    u_vec.apply("insert")
```

Comment:
So this is to re-assign the values of `u_new` to `u_cur`? Couldn't one just use `u_cur.assign(u_new)`?

Reply:
Yes, that would be easier :)